### ISSUE TITLE
Enable Cart Stock Reducer Issue

### ISSUE THREAD

**Issue by hobbyguts** [jamesgol/woocommerce-cart-stock-reducer#23]
Hi 
When i click "Enable Cart Stock Reducer", i found my stock level is correctly shown on the front page website. 
For example, my stock quantity of product A in the backend is 4, it shows only 3 at the frontend website.
Please clarify what's wrong with that

Thanks
Joe

**Comment by jamesgol** [comment]
The most common issue is those items are already added to existing carts.  From the backend you can go to WooCommerce->System Status->Tools and click 'Clear all sessions' to wipe out any existing sessions

**Comment by hobbyguts** [comment]
Thanks it works now!
My site is still at testing stage now but after launch, do i have to click "clear all sessions" regularly in order to make sure the quantity in frontend or backend matched please?

Besides, another question is there is total 2 quantity for an item, n when user A add 2 to the cart, then user B tries to add the same item to the cart, but the below message shows:
This item is not available at this time due to pending orders. Check back in 4 mins to see if items become available.
After 4 mins, user B tries to reload the page which still shows the same message, is that the cache problem please?

**Comment by jamesgol** [comment]
If you are using item expiration then you should not need to clear the sessions at all, since the items will expire at the configured time.

Caching could definitely cause an issue like that

**Comment by hobbyguts** [comment]
I did set the item expiration. Anyway let me try again later.
I found an interesting situation, when A user add 1 item to the cart, and before expiration time, he clicks to cancel this item at card, and at the same time B user manage to add this item to the add after A user cancellation.  However, A user changes his mind to click Undo to cancel the item, n he can also manage to add the item to the cart again.  
In this situation, the item only has 1 but both of them manage to add it to the cart.

**Comment by jamesgol** [comment]
Handling Undo is on the todo list, but I haven't had the time to work on that yet.  Most people hide or disable that functionality.

**Comment by hobbyguts** [comment]
OK Thanks.
I have one more question n m not sure if this is on the to do list also:
After proceed to paypal, the buyer will be directed to paypal page to settle payment, after finishing that, the buyer will be redirected to my website cart page.
However the below message is found as below;
Sorry, 'Item A"' was removed from your cart because you didn't checkout before the expiration time.
I think this is misleading as the buyer should already finish checkout and payment. After the buyer click my account section , this can shows the order processing.
Do you have any suggestion for this please?
Thanks

**Comment by hobbyguts** [comment]
Following up the above thread, perhaps due to this message " 'Item A"' was removed from your cart because you didn't checkout before the expiration time." , the website or other mobile website still record there is 1 item A in stock ; however, it should be 0 in inventory after the buyer finish the payment.

**Comment by jamesgol** [comment]
You need to make sure there is sufficient time available before expiration for them to checkout.  You could use one of WooCommerce's hooks to extend the expiration time when they hit the checkout if you want.

This is not a default setting because someone could then use paypal (or other methods) to infinitely hold an item in the cart.

**Comment by hobbyguts** [comment]
Okay, i got what you mean, the buyer must finish the whole checkout (finish paypal payment n go back to website) within the set Expire Time limit.  Then the above message will not be shown,
However, if I have set Hold Stock (minutes) , will it conflict with the expire time limit please? sorry i m a bit confusing 

Finally, you said to use one of WooCommerce's hooks to extend the expiration time when they hit the checkout , which function/plugin are you referring to please?

**Comment by jamesgol** [comment]
I actually forgot that I added something to help with this in a recent release. You can do is setup a filter on 'wc_csr_expire_ignore_status' that returns an array of WooCommerce status to keep from expiring.  I believe the paypal gateway sets the status to 'on-hold' when someone gets diverted to paypal.

**Comment by hobbyguts** [comment]
Sorry i m not a programmer ..so maybe i m waiting for your next update release for this issue. Thanks again for your reply

**Comment by jamesgol** [comment]
That feature is available now, with version 1.50 that came out 16 days ago.

**Comment by hobbyguts** [comment]
my version is 1.55 already but i could not see this setting: setup a filter on 'wc_csr_expire_ignore_status', where can i find that please? thx

**Comment by jamesgol** [comment]
This is something you would have to add to your themes functions.php or some other place you can add code.

**Comment by jamesgol** [comment]
I just added a configuration option to the settings page to allow you to select which order status is ignored.  That filter still gets called afterwards so you can still alter it via code.

If you want to try that version out you can download it from https://github.com/jamesgol/woocommerce-cart-stock-reducer/archive/master.zip

**Comment by jamesgol** [comment]
I also just committed code to change the Undo URL if the item is managed.  Now if you remove an item that is managed stock the URL returned will be the product page instead of a link that adds the item back to the cart.

**Comment by hobbyguts** [comment]
thanks a lot! I will try it out later!
In fact, i m just aware that what you mean a filter on 'wc_csr_expire_ignore_status' t is that like the buyer click "proceed to pal" button, this should trigger to check if there is any stock holding time, if yes, the expiry time should change to that holding time. 
However, i just imagine that if the buyer don't buy immediately at paypal checkout page, n come back to website later within holding time, then click "proceed to pal", would that holding time be recounted again?

**Comment by jamesgol** [comment]
It is a difficult situation using an external processor, generally there is no time limit once people get kicked over to the external processor.  It would be really nice if the processors supported some type of expiration time but I've never seen one that does.  Someone could always continually refresh their cart page on paypal to keep it alive.

What happens on the WooCommerce side when someone clicks the button to checkout with paypal is the order is added but the order status gets set to 'Pending' and the items remain in the users cart.  Without my plugin the items are available to purchase until they are actually paid for.  The big problem to avoid is people trying to buy items that there are only one quantity of, in a typical WooCommerce setup the item quantity is only reduced when the item is actually paid for this plugin attempts to minimize that time by making an item virtually unavailable when someone adds it to their cart.

If you have more specific needs I often have paid consulting time available.
